acceptancetests/assess_upgrade.py

```diff
@@ -0,0 +1,199 @@
+            devel_client, base_dir, 'proposed')
+        with stream_server.server() as url:
+            stable_client.env.update_config({
+                'agent-metadata-url': url,
+                'agent-stream': 'proposed'
+            })
+            with stable_bsm.booted_context(False):
+                assert_stable_model_is_correct(stable_client)
+
+                deploy_mediawiki_with_db(stable_client)
+                assert_media_wiki_is_responding(stable_client)
+                upgrade_stable_to_devel_version(devel_client)
+                assert_media_wiki_is_responding(devel_client)
+
+
+def upgrade_stable_to_devel_version(client):
+    devel_version = get_stripped_version_number(client.version)
+    client.get_controller_client().juju('upgrade-juju', ('-m', 'controller'))
+    assert_model_is_version(client.get_controller_client(), devel_version)
+    wait_until_model_upgrades(client)
+
+    client.juju('upgrade-juju', ())
+    assert_model_is_version(client, devel_version)
+    wait_until_model_upgrades(client)
+
+
+def wait_until_model_upgrades(client, timeout=300):
+    # Poll using a command that will fail until the upgrade is complete.
+    def model_upgrade_status_check(client):
+        try:
+            log.info('Attempting API connection, failure is not fatal.')
+            client.list_models()
+            client.juju('list-users', (), include_e=False)
+            return True
+        except CalledProcessError:
+            # Upgrade will still be in progress and thus refuse the api call.
+            return False
+    try:
+        wait_for_model_check(client, model_upgrade_status_check, timeout)
+    except ModelCheckFailed:
+        raise JujuAssertionError(
+            'Upgrade for model {} failed to complete within the alloted '
+            'timeout ({} seconds)'.format(
+                client.model_name, timeout))
+
+
+def assert_stable_model_is_correct(stable_client):
+    assert_model_is_version(
+        stable_client.get_controller_client(),
+        get_stripped_version_number(stable_client.version))
+    assert_model_is_version(
+        stable_client,
+        get_stripped_version_number(stable_client.version))
+
+
+def assert_model_is_version(client, expected_version, timeout=600):
+    # Check agent & machine version
+    client.wait_for_version(expected_version, timeout)
+    # Also check model version
+    client.wait_for(WaitModelVersion(expected_version, timeout))
+
+
+def setup_agent_metadata(stream_server, agent, client, tmp_dir, stream):
+    version_parts = get_version_parts(client.version)
+
+    if agent is None:
+        agent_details = agent_tgz_from_juju_binary(client.full_path, tmp_dir)
+    else:
+        agent_details = agent
+
+    stream_server.add_product(
+        stream,
+        version_parts.version,
+        version_parts.arch,
+        client.env.get_option('default-series'),
+        agent_details)
+    # Trusty needed for wikimedia charm.
+    stream_server.add_product(
```

Comment:
Ugh, hardcoding trusty. Fun :-) Does wikimedia REALLY need trusty, or is the charm metadata just old?

Reply:
I'm trying a run with it being forced to xenial. Even if it does work, the charm 'supports' trusty, do we want to run the risk of failure because we're using xenial instead? (so not an actual juju error)

Reply:
Yeah, wikimedia charm errors if forced to use xenial.

Reply:
Ok, we tried :-)